arch/x86/cpu/tangier/pinmux.c

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) 2018 Emlid Limited
+ */
+
+#include <common.h>
+#include <dm.h>
+#include <regmap.h>
+#include <syscon.h>
+#include <asm/cpu.h>
+#include <asm/scu.h>
+#include <linux/io.h>
+
+#define BUFCFG_OFFSET	0x100
+
+#define MRFLD_FAMILY_LEN	0x400
+
+#define pin_to_bufno(f, p)	  ((p) - (f)->pin_base)
+
+struct mrfld_family {
+	unsigned int family_number;
+	unsigned int pin_base;
+	size_t npins;
+	void __iomem *regs;
+};
+
+#define MRFLD_FAMILY(b, s, e)				\
+	{				\
+		.family_number = (b),			\
+		.pin_base = (s),			\
+		.npins = (e) - (s) + 1,			\
+	}
+
+static struct mrfld_family i2c_family = MRFLD_FAMILY(7, 101, 114);
+
+static void __iomem *mrfld_get_bufcfg(struct mrfld_family *family, unsigned int pin)
+{
+	unsigned int bufno;
+
+	bufno = pin_to_bufno(family, pin);
+
+	return family->regs + BUFCFG_OFFSET + bufno * 4;
+}
+
+static void mrfld_setup_families(void *base_addr, struct mrfld_family *families, unsigned int nfam)
+{
+    for (int i = 0; i < nfam; i++) {
+		struct mrfld_family *family = &families[i];
+
+		family->regs = base_addr + family->family_number * MRFLD_FAMILY_LEN;
+	}
+}
+
+int mrfld_pinconfig_protected(unsigned int pin, u32 mask, u32 bits)
```

Comment:
Since this is an API it has to get a pin control device first, something like

        ret = syscon_get_by_driver_data(X86_SYSCON_SCU, &dev);

The idea behind is to get lazy probe (that's called "Driver Model" in U-Boot).

Reply:
@andy-shev 
Thanks. Makes sense.

But it seems like probe is already automagically called just by the sheer addition of the probe callback, doesn't it? I'm pretty sure I'm missing something, of course.

Reply:
If you call function before device is probed, it wouldn't work and crash everything. ->probe() will be called when we try to get pointer to a device.